cmd/zfs/zfs_project.c

```diff
@@ -0,0 +1,444 @@
+/*
+ * CDDL HEADER START
+ *
+ * The contents of this file are subject to the terms of the
+ * Common Development and Distribution License (the "License").
+ * You may not use this file except in compliance with the License.
+ *
+ * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+ * or http://www.opensolaris.org/os/licensing.
+ * See the License for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+
+/*
+ * Copyright (c) 2017, Intle Corporation. All rights reserved.
+ */
+
+#include <errno.h>
+#include <getopt.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <strings.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <dirent.h>
+#include <stddef.h>
+#include <libintl.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/list.h>
+#include <sys/zfs_project.h>
+
+#include "zfs_util.h"
+#include "zfs_projectutil.h"
+
+typedef struct zfs_project_item {
+	list_node_t	zpi_list;
+	char		zpi_pathname[PATH_MAX];
+} zfs_project_item_t;
+
+static int
+zfs_project_item_alloc(list_t *head, const char *pathname)
+{
+	zfs_project_item_t *zpi;
+
+	zpi = safe_malloc(sizeof (zfs_project_item_t));
+	if (zpi == NULL) {
+		(void) fprintf(stderr, gettext("not enough RAM to handle %s\n"),
+		    pathname);
+		return (-1);
+	}
+
+	strcpy(zpi->zpi_pathname, pathname);
+	list_insert_tail(head, zpi);
+
+	return (0);
+}
+
+static int
+zfs_project_handle_one(const char *pathname, uint64_t projid,
+    zfs_project_ops_t op, int hint)
+{
+	zfsxattr_t fsx;
+	int ret, fd;
+
+	fd = open(pathname, O_RDONLY | O_NOCTTY);
```

Comment:
This will follow symlinks and potentially set the project ID and flags on files outside of the project tree. I'm not sure of the best way to handle that. Using `O_NOFOLLOW | O_PATH` returns a file descriptor for the symlink itself, but using it with ioctl() returns EBADF. We could just ignore symlinks (this is what xfs_quota does) but then objects consumed by symbolic links won't be accounted for. Maybe that's a reasonable compromise as most filesystems hopefully contain relatively few symlinks. :thinking:

Reply:
Yes, we definitely don’t want to be changing the targets of the symlinks. Either they are outside the tree and should not get the projid set, or they are inside the tree and will be updated directly.

Reply:
We should keep the behavior regarding symlinks consistent in order to avoid any confusion.  Since it's problematic to set project ID on the symlink itself with `zfs project`, let's ignore them which has the advantage of being consistent with xfs.  We'll also need to make the same change on the kernel side to ignore newly created symlinks in a managed directory.  We don't want some to be accounted and not others.

We should add a test case for symlinks and another to verify the other special file types work correctly (char device file, block device file, local socket, and a named pipe).  If they end up being problematic they can be ignored too.